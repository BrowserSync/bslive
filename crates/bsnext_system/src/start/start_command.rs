#[derive(Debug, Clone, clap::Parser)]
pub struct StartCommand {
    /// Should permissive cors headers be added to all responses?
    #[arg(long)]
    pub cors: bool,

    /// Specify a port instead of a random one
    #[arg(short, long)]
    pub port: Option<u16>,

    /// Paths to serve + possibly watch, incompatible with `-i` option
    pub trailing: Vec<String>,
}
