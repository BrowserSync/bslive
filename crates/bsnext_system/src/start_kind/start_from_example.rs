use bsnext_example::Example;
use bsnext_input::server_config::Identity;
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::{fs_write_input, rand_word, DirError, InputError};
use std::fs;

#[derive(Debug)]
pub struct StartFromExample {
    pub example: Example,
    pub write_input: bool,
    pub target_kind: TargetKind,
    pub port: Option<u16>,
    pub temp: bool,
    pub name: Option<String>,
}

impl SystemStart for StartFromExample {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        let identity = Identity::from_port_or_named(self.port)?;
        let input = self.example.into_input(identity);
        let name = self.name.clone();
        let dir = if self.temp {
            let temp_dir = std::env::temp_dir();
            let word = name.unwrap_or_else(rand_word);
            let num = rand::random::<f64>();
            let next_dir = temp_dir.join(format!("bslive-{word}-{num}"));
            fs::create_dir_all(&next_dir)
                .map_err(|_e| DirError::CannotCreate {
                    path: next_dir.clone(),
                })
                .and_then(|_pb| {
                    std::env::set_current_dir(&next_dir).map_err(|_e| DirError::CannotMove {
                        path: next_dir.clone(),
                    })
                })
                .map(|_| next_dir.clone())?
        } else {
            ctx.cwd.to_path_buf()
        };
        if self.write_input {
            let path = fs_write_input(&dir, &input, self.target_kind.clone())?;
            Ok(SystemStartArgs::PathWithInput { path, input })
        } else {
            Ok(SystemStartArgs::InputOnly { input })
        }
    }
}
