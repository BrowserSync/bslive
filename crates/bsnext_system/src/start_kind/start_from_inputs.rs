use crate::startup::{StartupContext, SystemStart, SystemStartArgs};

use bsnext_input::{Input, InputError};
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct StartFromInputPaths {
    pub input_paths: Vec<String>,
}

impl SystemStart for StartFromInputPaths {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        from_yml_paths(&ctx.cwd, &self.input_paths)
    }
}

fn from_yml_paths<T: AsRef<str>>(cwd: &Path, inputs: &[T]) -> Result<SystemStartArgs, InputError> {
    let input_candidates = inputs
        .iter()
        .map(|path| cwd.join(path.as_ref()))
        .collect::<Vec<PathBuf>>();

    let lookups = ["input.yml", "input.yaml"]
        .iter()
        .map(|path| cwd.join(path))
        .collect::<Vec<PathBuf>>();

    let auto_candidates = lookups
        .iter()
        .filter(|path| Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    let exists = input_candidates
        .iter()
        .filter(|path| Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    let missing = input_candidates
        .iter()
        .filter(|path| !Path::exists(path))
        .collect::<Vec<&PathBuf>>();

    if !missing.is_empty() {
        for path in &missing {
            tracing::error!(?path, "input file not found");
        }
        return Err(InputError::NotFound(
            missing.first().expect("guarded").to_path_buf(),
        ));
    }

    let first_user = exists.first();
    let first_auto = auto_candidates.first();

    let Some(input_path) = first_user.or(first_auto) else {
        return Err(InputError::MissingInputs);
    };

    tracing::info!(?input_path);

    let result = Input::from_input_path(input_path);
    match result {
        Ok(input) => Ok(SystemStartArgs::PathWithInput {
            path: input_path.to_path_buf(),
            input,
        }),
        Err(InputError::YamlError(yaml_error)) => Ok(SystemStartArgs::PathWithInvalidInput {
            path: input_path.to_path_buf(),
            input_error: InputError::YamlError(yaml_error),
        }),
        Err(e) => {
            tracing::error!("cannot continue");
            Err(e)
        }
    }
}
