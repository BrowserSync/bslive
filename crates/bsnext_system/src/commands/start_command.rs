use crate::args::Args;
use crate::start_kind::StartKind;
use crate::{BsSystem, Start};
use actix::Actor;
use bsnext_dto::internal::{AnyEvent, StartupEvent};
use bsnext_input::startup::{DidStart, StartupError};
use bsnext_output::stdout::StdoutTarget;
use bsnext_output::{OutputWriterTrait, OutputWriters};
use bsnext_tracing::OutputFormat;
use std::path::PathBuf;
use tokio::sync::{mpsc, oneshot};
use tracing::debug_span;

pub async fn start_cmd(cwd: PathBuf, args: Args) -> Result<(), impl OutputWriterTrait> {
    let (tx, rx) = oneshot::channel();
    let (events_sender, mut events_receiver) = mpsc::channel::<AnyEvent>(1);

    let system = BsSystem::new();
    let sys_addr = system.start();
    let format_clone = args.format;
    let start_kind = StartKind::from_args(&args);

    tracing::debug!(?start_kind);

    let start = Start {
        kind: start_kind,
        cwd: Some(cwd),
        ack: tx,
        events_sender,
    };

    match sys_addr.send(start).await {
        Ok(Ok(DidStart::Started)) => {
            // everything good here, continue...
        }
        Ok(Err(e)) => {
            return Err(StartupEvent::FailedStartup(e));
        }
        Err(e) => {
            let message = e.to_string();
            return Err(StartupEvent::FailedStartup(StartupError::Other(message)));
        }
    };

    let events_handler = tokio::spawn(async move {
        // at this point, we started, so we can choose a TUI
        let printer = match format_clone {
            OutputFormat::Tui => {
                // let rr = Ratatui::try_new().expect("test");
                // let (sender, _ui_handle, _other) = rr.install().expect("thread install");
                // Writers::Ratatui(sender)
                todo!("re-implement ratatui")
            }
            OutputFormat::Json => OutputWriters::Json,
            OutputFormat::Normal => OutputWriters::Pretty,
        };
        // let events = vec![];
        let stdout = &mut std::io::stdout();
        let stderr = &mut std::io::stderr();
        let mut sink = StdoutTarget::new(stdout, stderr);

        while let Some(evt) = events_receiver.recv().await {
            let span = debug_span!("External Event processor");
            let _g2 = span.enter();
            tracing::debug!(external_event = ?evt);
            let result = match evt {
                AnyEvent::Internal(int) => printer.write_evt(&int, &mut sink.output()),
                AnyEvent::External(ext) => printer.write_evt(&ext, &mut sink.output()),
            };
            match result {
                Ok(_) => {}
                Err(_) => tracing::error!("could not handle event"),
            }
            sink.flush();
        }
    });

    match rx.await {
        Ok(_) => {
            tracing::info!("servers ended");
        }
        Err(e) => {
            // dropped? this is ok
            tracing::trace!(?e, "");
        }
    };

    match events_handler.await {
        Ok(_) => {
            tracing::info!("events ended");
        }
        Err(e) => {
            // dropped? this is ok
            tracing::trace!(?e, "");
        }
    }

    Ok(())
}
