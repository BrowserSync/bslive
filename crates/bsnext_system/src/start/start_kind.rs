use crate::start::start_command::StartCommand;
use crate::start::start_kind::start_from_example::StartFromExample;
use crate::start::start_kind::start_from_inputs::{StartFromInput, StartFromInputPaths};
use crate::start::start_kind::start_from_paths::StartFromTrailingArgs;
use bsnext_core::shared_args::{FsOpts, InputOpts};
use bsnext_fs_helpers::{fs_write_str, FsWriteError, WriteMode};
use bsnext_input::route::{CorsOpts, Opts, Route};
use bsnext_input::server_config::{ServerConfig, ServerIdentity};
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::InputWriter;
use bsnext_input::{Input, InputError};
use std::path::{Path, PathBuf};

pub mod start_from_example;
pub mod start_from_inputs;
pub mod start_from_paths;

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
pub enum StartKind {
    FromInput(StartFromInput),
    FromInputPaths(StartFromInputPaths),
    FromExample(StartFromExample),
    FromTrailingArgs(StartFromTrailingArgs),
}

impl StartKind {
    pub fn from_args(fs_opts: &FsOpts, input_opts: &InputOpts, cmd: &StartCommand) -> Self {
        // todo: re-implement example command
        // if let Some(example) = args.example {
        //     return StartKind::FromExample(StartFromExample {
        //         example,
        //         write_input: args.write,
        //         port: args.port,
        //         temp: args.temp,
        //         name: args.name.clone(),
        //         target_kind: args
        //             .target
        //             .as_ref()
        //             .map(ToOwned::to_owned)
        //             .unwrap_or_default(),
        //         dir: args.dir.clone(),
        //         force: args.force,
        //     });
        // }

        // todo: make the addition of a proxy + route opts easier?
        if cmd.trailing.is_empty() {
            tracing::debug!("0 trailing, {} inputs", input_opts.input.len());
            if input_opts.input.is_empty() && !cmd.proxies.is_empty() {
                let first_proxy = cmd.proxies.first().expect("guarded first proxy");
                let r = Route::proxy(first_proxy);
                let id = ServerIdentity::from_port_or_named(cmd.port).unwrap_or_else(|_e| {
                    tracing::error!("A problem occurred with the port?");
                    ServerIdentity::named()
                });
                let ser = ServerConfig::from_route(r, id);
                let input = Input::from_server(ser);
                StartKind::FromInput(StartFromInput { input })
            } else {
                StartKind::FromInputPaths(StartFromInputPaths {
                    input_paths: input_opts.input.clone(),
                    port: cmd.port,
                })
            }
        } else {
            StartKind::FromTrailingArgs(StartFromTrailingArgs {
                paths: cmd.trailing.clone(),
                write_input: fs_opts.write,
                port: cmd.port,
                force: fs_opts.force,
                route_opts: Opts {
                    cors: cmd.cors.then_some(CorsOpts::Cors(true)),
                    ..Default::default()
                },
            })
        }
    }
    pub fn from_input(input: Input) -> Self {
        Self::FromInput(StartFromInput { input })
    }
}

impl SystemStart for StartKind {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, Box<InputError>> {
        match self {
            StartKind::FromInputPaths(from_inputs) => from_inputs.input(ctx),
            StartKind::FromExample(from_example) => from_example.input(ctx),
            StartKind::FromTrailingArgs(from_trailing_args) => from_trailing_args.input(ctx),
            StartKind::FromInput(from_inputs) => from_inputs.input(ctx),
        }
    }
}

pub fn fs_write_input(
    cwd: &Path,
    input: &Input,
    target_kind: TargetKind,
    write_mode: &WriteMode,
) -> Result<PathBuf, FsWriteError> {
    let string = match target_kind {
        TargetKind::Yaml => bsnext_yaml::yaml_writer::YamlWriter.input_to_str(input),
        TargetKind::Toml => todo!("toml missing"),
        TargetKind::Md => bsnext_md::md_writer::MdWriter.input_to_str(input),
        TargetKind::Html => bsnext_html::html_writer::HtmlWriter.input_to_str(input),
    };
    let name = match target_kind {
        TargetKind::Yaml => "bslive.yml",
        TargetKind::Toml => todo!("toml missing"),
        TargetKind::Md => "bslive.md",
        TargetKind::Html => "bslive.html",
    };

    fs_write_str(cwd, &PathBuf::from(name), &string, write_mode)
}
