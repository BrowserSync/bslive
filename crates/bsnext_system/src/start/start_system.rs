use crate::start::start_command::StartCommand;
use crate::start::start_kind::StartKind;
use crate::{BsSystem, Start};
use actix::{Actor, Addr};
use bsnext_core::shared_args::{FsOpts, InputOpts};
use bsnext_dto::internal::{AnyEvent, StartupEvent};
use bsnext_input::startup::{DidStart, StartupError};
use bsnext_output::OutputWriterTrait;
use std::path::PathBuf;
use tokio::sync::oneshot;

pub async fn start_system(
    cwd: PathBuf,
    fs_opts: FsOpts,
    input_opts: InputOpts,
    start_command: StartCommand,
    events_sender: tokio::sync::mpsc::Sender<AnyEvent>,
) -> Result<(oneshot::Receiver<()>, Addr<BsSystem>), impl OutputWriterTrait> {
    let (tx, rx) = oneshot::channel();
    let system = BsSystem::new();
    let sys_addr = system.start();
    let start_kind = StartKind::from_args(&fs_opts, &input_opts, &start_command);

    tracing::debug!(?start_kind);

    let start = Start {
        kind: start_kind,
        cwd: Some(cwd),
        ack: tx,
        events_sender,
    };

    match sys_addr.send(start).await {
        Ok(Ok(DidStart::Started)) => Ok((rx, sys_addr)),
        Ok(Err(e)) => Err(StartupEvent::FailedStartup(e)),
        Err(e) => {
            let message = e.to_string();
            Err(StartupEvent::FailedStartup(StartupError::Other(message)))
        }
    }
}
