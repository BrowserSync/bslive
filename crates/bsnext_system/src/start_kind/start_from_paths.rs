use crate::start_kind::start_fs;
use bsnext_fs_helpers::WriteMode;
use bsnext_input::route::{DirRoute, Route, RouteKind};
use bsnext_input::server_config::{ServerConfig, ServerIdentity};
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};
use bsnext_input::target::TargetKind;
use bsnext_input::{Input, InputError, PathDefinition, PathDefs, PathError};
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct StartFromDirPaths {
    pub paths: Vec<String>,
    pub write_input: bool,
    pub port: Option<u16>,
    pub force: bool,
}

impl SystemStart for StartFromDirPaths {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, Box<InputError>> {
        let identity =
            ServerIdentity::from_port_or_named(self.port).map_err(|e| Box::new(e.into()))?;
        let input =
            from_dir_paths(&ctx.cwd, &self.paths, identity).map_err(|e| Box::new(e.into()))?;
        let write_mode = if self.force {
            WriteMode::Override
        } else {
            WriteMode::Safe
        };
        if self.write_input {
            let path = start_fs::fs_write_input(&ctx.cwd, &input, TargetKind::Yaml, &write_mode)
                .map_err(|e| Box::new(e.into()))?;
            Ok(SystemStartArgs::PathWithInput { input, path })
        } else {
            Ok(SystemStartArgs::InputOnly { input })
        }
    }
}

fn from_dir_paths<T: AsRef<str>>(
    cwd: &Path,
    paths: &[T],
    identity: ServerIdentity,
) -> Result<Input, PathError> {
    let path_defs = paths
        .iter()
        .map(|p| {
            let pb = PathBuf::from(p.as_ref());
            if pb.is_absolute() {
                return PathDefinition {
                    input: p.as_ref().to_string(),
                    cwd: cwd.to_path_buf(),
                    absolute: pb,
                };
            } else {
                return PathDefinition {
                    input: p.as_ref().to_string(),
                    cwd: cwd.to_path_buf(),
                    absolute: cwd.join(pb),
                };
            }
        })
        .map(|path_def| {
            let exists = path_def.absolute.exists();
            (path_def, exists)
        })
        .collect::<Vec<(PathDefinition, bool)>>();

    let invalid = path_defs
        .into_iter()
        .filter_map(|(pb, exists)| if exists { None } else { Some(pb) })
        .collect::<Vec<_>>();

    if !invalid.is_empty() {
        return Err(PathError::MissingPaths {
            paths: PathDefs(invalid),
        });
    }

    let server = ServerConfig {
        identity,
        routes: paths
            .iter()
            .map(|p| -> Route {
                let str = p.as_ref();
                Route {
                    path: "/".to_string().parse().unwrap(),
                    kind: RouteKind::Dir(DirRoute {
                        dir: str.into(),
                        base: None,
                    }),
                    ..Default::default()
                }
            })
            .collect(),
        ..Default::default()
    };
    Ok(Input {
        servers: vec![server],
    })
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test() -> anyhow::Result<()> {
        use tempfile::tempdir;
        let tmp_dir = tempdir()?;
        let v = StartFromDirPaths {
            paths: vec![".".into()],
            write_input: false,
            port: Some(3000),
            force: false,
        };
        let ctx = StartupContext {
            cwd: tmp_dir.path().to_path_buf(),
        };
        let i = v.input(&ctx);
        tmp_dir.close()?;
        let start_args = i.unwrap();
        if let SystemStartArgs::InputOnly { input } = start_args {
            insta::assert_debug_snapshot!(input);
            insta::assert_yaml_snapshot!(input);
        } else {
            unreachable!("cannot get here?")
        }
        Ok(())
    }
}
