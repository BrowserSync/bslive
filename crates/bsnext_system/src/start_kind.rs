use crate::args::Args;
use crate::start_kind::start_from_example::StartFromExample;
use crate::start_kind::start_from_inputs::{StartFromInput, StartFromInputPaths};
use crate::start_kind::start_from_paths::StartFromPaths;
use bsnext_input::startup::{StartupContext, SystemStart, SystemStartArgs};

use bsnext_input::{Input, InputError};

pub mod start_from_example;
pub mod start_from_inputs;
pub mod start_from_paths;

#[derive(Debug)]
pub enum StartKind {
    FromInput(StartFromInput),
    FromInputPaths(StartFromInputPaths),
    FromExample(StartFromExample),
    FromPaths(StartFromPaths),
}

impl StartKind {
    pub fn from_args(args: Args) -> Self {
        if let Some(example) = args.example {
            return StartKind::FromExample(StartFromExample {
                example,
                write_input: args.write,
                port: args.port,
                temp: args.temp,
                name: args.name,
                target_kind: args.target.unwrap_or_default(),
            });
        }

        if !args.paths.is_empty() {
            StartKind::FromPaths(StartFromPaths {
                paths: args.paths,
                write_input: args.write,
                port: args.port,
            })
        } else {
            StartKind::FromInputPaths(StartFromInputPaths {
                input_paths: args.input.clone(),
            })
        }
    }
    pub fn from_input(input: Input) -> Self {
        Self::FromInput(StartFromInput { input })
    }
}

impl SystemStart for StartKind {
    fn input(&self, ctx: &StartupContext) -> Result<SystemStartArgs, InputError> {
        match self {
            StartKind::FromInputPaths(from_inputs) => from_inputs.input(ctx),
            StartKind::FromExample(from_example) => from_example.input(ctx),
            StartKind::FromPaths(from_input_paths) => from_input_paths.input(ctx),
            StartKind::FromInput(from_inputs) => from_inputs.input(ctx),
        }
    }
}
