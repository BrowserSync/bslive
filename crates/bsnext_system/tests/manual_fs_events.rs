use bsnext_core::shared_args::{FsOpts, InputOpts};
use bsnext_dto::external_events::ExternalEventsDTO;
use bsnext_dto::internal::AnyEvent;
use bsnext_fs::FsEvent;
use bsnext_system::start::start_command::StartCommand;
use bsnext_system::start::start_kind::StartKind;
use bsnext_system::start::start_system::start_system;
use futures_util::future::join;
use futures_util::StreamExt;
use std::fs;
use std::path::PathBuf;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;

#[actix_rt::test]
async fn test_fs_events() -> Result<(), anyhow::Error> {
    let tmp_dir = tempfile::tempdir().unwrap();
    let index_file = tmp_dir.path().join("bslive.yaml");
    let input = r#"
servers:
    - name: api
      routes:
        - path: /
          html: |
            <body>
              <link href="/c.css" rel="stylesheet">
            </body>
          watch:
            run:
              - sh: echo 'oops'
        - path: /c.css
          raw: "body { background: cyan }"
    "#;
    fs::write(&index_file, input).expect("can write?");

    let cwd = PathBuf::from(tmp_dir.path());

    let (events_sender, events_receiver) = mpsc::channel::<AnyEvent>(1);
    let start = StartCommand {
        cors: false,
        port: None,
        proxies: vec![],
        trailing: vec![],
        logging: Default::default(),
        format: Default::default(),
    };
    let start_kind = StartKind::from_args(&FsOpts::default(), &InputOpts::default(), &start);
    let api = start_system(cwd, start_kind, events_sender)
        .await
        .map_err(|e| anyhow::anyhow!("{:?}", e))?;

    let active_servers = api.active_servers().await?;
    let first = active_servers.first().unwrap();
    let id = first.identity.as_id();

    let r = ReceiverStream::new(events_receiver)
        // .inspect(|e| println!("{:#?}", e))
        .filter(|e| {
            futures::future::ready(matches!(
                &e,
                AnyEvent::External(ExternalEventsDTO::FileChanged(..))
            ))
        })
        .take(2)
        .collect::<Vec<_>>();

    let h1 = tokio::spawn(r);

    let h = tokio::spawn(async move {
        api.fs_event(FsEvent::changed("/c.css", "c.css", id));
        tokio::time::sleep(Duration::from_millis(10)).await;
        api.fs_event(FsEvent::changed("/c.css", "c.css", id));
    });

    let (events, _) = join(h1, h).await;
    insta::assert_debug_snapshot!(events);
    Ok(())
}
