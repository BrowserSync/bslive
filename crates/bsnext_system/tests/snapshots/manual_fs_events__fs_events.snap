---
source: crates/bsnext_system/tests/manual_fs_events.rs
expression: events
---
Ok(
    [
        External(
            FileChanged(
                FileChangedDTO {
                    path: "c.css",
                },
            ),
        ),
        External(
            FileChanged(
                FileChangedDTO {
                    path: "c.css",
                },
            ),
        ),
    ],
)
